### PR TITLE
feat!: store both recurring/one-time grants & spents; add `state`

### PR DESCRIPTION
**Pull request by sidvishnoi**
## Context

Closes https://github.com/interledger/web-monetization-extension/issues/376

## Changes proposed in this pull request

1. Update storage as suggested in https://github.com/interledger/web-monetization-extension/issues/376, but with following differences:
   - Instead of storing `interval` and `approvedAmount` as separate keys in grants, store them as `amount: WalletAmount` (but within the grant)
   - Instead of storing balances, we'll store spent amounts as API only returns spent amounts.
   - Add `GrantDetails.type = "recurring" | "one-time"` to make it easier for TypeScript to discriminate between two.
2. Store a `Storage.version` to help with migrations, keep migrations indexed by this sequential version number.
3. While we're doing a migration here:
   - Add `Storage.state`. This is also a precursor to https://github.com/interledger/web-monetization-extension/issues/144 (`out_of_funds` state)
   - Replace `Storage.hasHostPermissions` with `Storage.state = "missing_host_permissions"` 

Also includes some parts of https://github.com/interledger/web-monetization-extension/issues/144 (setting the first available grant, but without `switchGrant`) and https://github.com/interledger/web-monetization-extension/issues/272.

### CODE REVIEW

**Comment by raducristianpopa** [comment]
<h2>Extension builds preview</h2>

<table role="table">
<thead>
  <tr>
    <th>Name</th>
    <th>Link</th>
  </tr>
</thead>
<tbody>
  <tr>
    <td align="center">Latest commit</td>
    <td align="center">0514f88aa10a0e2ba358734b9bf092812d825e52</td>
  </tr>
  <tr>
    <td align="center">Latest job logs</td>
    <td align="center"><a href="https://github.com/interledger/web-monetization-extension/actions/runs/9711016075">Run #9711016075</a></td>
  </tr>
  <tr><td align="center">Badge</td><td align="center">N/A</td></tr><tr><td align="center">Badge</td><td align="center">N/A</td></tr>
</tbody>
</table>

**Comment by sidvishnoi** [comment]
Before the `Storage.state` commit, changed storage like this, seamlessly on extension update:

Screenshot from 2024-06-28 14-33-34

File: src/background/services/openPayments.ts [hunk 1]
```diff
@@ -330,25 +338,41 @@ export class OpenPaymentsService {
...
+        recurringGrant: grantDetails,
+        recurringGrantRemainingBalance: amount
+      })
+    } else {
+      await this.storage.set({
+        ...data,
+        oneTimeGrant: grantDetails,
+        oneTimeGrantRemainingBalance: amount
+      })
+    }
```
Comment (sidvishnoi) [inline src/background/services/openPayments.ts:373]: Nit: Had to be verbose here as TypeScript was causing issues with 
```ts
const data = {
  // ... common properties
  ...(type === "recurring" ? {recurringGrant: grantDetails} : {oneTimeGrant: grantDetails})
```

File: src/background/services/openPayments.ts [hunk 2]
```diff
@@ -330,25 +338,41 @@ export class OpenPaymentsService {
...
+    this.grant = grantDetails
+    this.token = this.grant.accessToken
```
Comment (sidvishnoi) [inline src/background/services/openPayments.ts:375]: Nit: Keeping `this.token` to avoid unrelated changes (we can use this.grant.accessToken` everywhere now).

File: src/background/services/storage.ts [hunk 3]
```diff
@@ -1,16 +1,23 @@
...
+  GrantDetails,
+  Storage,
+  StorageKey,
+  WalletAmount
+} from '@/shared/types'
 import { type Browser } from 'webextension-polyfill'
 import { EventsService } from './events'
 
 const defaultStorage = {
+  version: 2,
```
Comment (sidvishnoi) [inline src/background/services/storage.ts:11]: For migrations, change this version and add a migration script in `MIGRATIONS` below.
New additions that can be dynamically set don't need migrations (e.g. we check if value is null etc.) but other structural changes would need migrations for keeping compatibility with existing installations. Also see the comment in `MIGRATIONS`.

Comment (sidvishnoi) [reply src/background/services/storage.ts:11]: I'll add this comment in code.

File: src/shared/types.ts [hunk 4]
```diff
@@ -13,41 +13,60 @@ export interface Amount {
...
+  amount: Omit<WalletAmount, 'interval'>
+}
+export interface RecurringGrant extends GrantDetailsBase {
+  type: 'recurring'
+  amount: Required<WalletAmount>
 }
+export type GrantDetails = OneTimeGrant | RecurringGrant
+
+/** Bigint amount, before transformation with assetScale */
+type AmountValue = string
```
Comment (raducristianpopa) [inline src/shared/types.ts:37]: Let's use this for `WalletAmount['value']` and `Amount['value']` as well.

File: package.json [hunk 5]
```diff
@@ -104,7 +104,7 @@
     "pnpm": "^9.4.0",
     "npm": "pnpm",
     "yarn": "pnpm",
-    "node": "20.14.0"
+    "node": "^20.14"
```
Comment (raducristianpopa) [inline package.json:107]: See my comment in #380.

File: src/background/services/background.ts [hunk 6]
```diff
@@ -154,28 +155,37 @@ export class Background {
...
     })
   }
 
   checkPermissions = async () => {
     try {
       this.logger.debug('checking hosts permission')
-      const status = await this.browser.permissions.contains(PERMISSION_HOSTS)
-      this.storage.setHostPermissionStatus(status)
+      const ok = await this.browser.permissions.contains(PERMISSION_HOSTS)
+      this.storage.setState({ missing_host_permissions: !ok })
```
Comment (raducristianpopa) [inline src/background/services/background.ts:188]:
```suggestion
      const hasPermissions = await this.browser.permissions.contains(PERMISSION_HOSTS)
      this.storage.setState({ missing_host_permissions: !hasPermissions })
```

File: src/background/services/events.ts [hunk 7]
```diff
@@ -1,8 +1,12 @@
 import { EventEmitter } from 'events'
+import type { Storage } from '@/shared/types'
 
 interface BackgroundEvents {
   'storage.rate_of_pay_update': { rate: string }
-  'storage.host_permissions_update': { status: boolean }
+  'storage.state_update': {
+    state: Storage['state']
+    prevState: Storage['state']
+  }
```
Comment (raducristianpopa) [inline src/background/services/events.ts:9]: Why are we passing the current state and the previous state?

File: src/background/services/background.ts [hunk 8]
```diff
@@ -154,28 +155,37 @@ export class Background {
...
+      const existingData = await this.storage.get()
+      this.logger.info(existingData)
       if (details.reason === 'install') {
         await this.storage.populate()
         await this.openPaymentsService.generateKeys()
       }
+      const migrated = await this.storage.migrate()
+      if (migrated) {
+        const prevVersion = existingData.version ?? 1
+        this.logger.info(`Migrated from ${prevVersion} to ${migrated.version}`)
```
Comment (raducristianpopa) [inline src/background/services/background.ts:178]: I know we are explicitly checking for the version in the `migrate` method to prevent migrating multiple times, but let's check for `details.reason === 'update'` as well.

File: src/background/services/background.ts [hunk 9]
```diff
@@ -154,28 +155,37 @@ export class Background {
...
-      this.logger.info(await this.storage.get())
+      const existingData = await this.storage.get()
+      this.logger.info(existingData)
       if (details.reason === 'install') {
         await this.storage.populate()
         await this.openPaymentsService.generateKeys()
       }
+      const migrated = await this.storage.migrate()
+      if (migrated) {
+        const prevVersion = existingData.version ?? 1
```
Comment (raducristianpopa) [inline src/background/services/background.ts:177]: Opinions on setting the storage version with Webpack defines?

File: src/background/services/background.ts [hunk 10]
```diff
@@ -154,28 +155,37 @@ export class Background {
...
+    this.events.on('storage.state_update', async ({ state, prevState }) => {
+      this.logger.info('state changed', { state, prevState })
       // TODO: change icon here in future
     })
   }
 
   bindOnInstalled() {
     this.browser.runtime.onInstalled.addListener(async (details) => {
-      this.logger.info(await this.storage.get())
+      const existingData = await this.storage.get()
```
Comment (raducristianpopa) [inline src/background/services/background.ts:169]:
```suggestion
      const data = await this.storage.get()
```

File: src/background/services/openPayments.ts [hunk 11]
```diff
@@ -459,18 +483,17 @@ export class OpenPaymentsService {
   }
 
   async disconnectWallet() {
-    const { grant } = await this.storage.get(['grant'])
+    const { recurringGrant, oneTimeGrant } = await this.storage.get([
+      'recurringGrant',
+      'oneTimeGrant'
+    ])
+    const grant = recurringGrant || oneTimeGrant
```
Comment (raducristianpopa) [inline src/background/services/openPayments.ts:490]: We should try to revoke both grants of wallet disconnect. For example:
- User sets a recurring amount when connecting the wallet
- Runs out of funds and adds a one-time budget
- User disconnects the wallet.

In this case, we will only revoke the recurring grant, but not the one-time one.

File: src/background/services/background.ts (hunk 9 above)
Comment (sidvishnoi) [reply src/background/services/background.ts:177]: It's in defaultStorage as we'll update version typically when we update values in defaultStorage (or its structure). Having it in defines won't make sense here, as it's a fixed part of code.

File: src/background/services/openPayments.ts (hunk 11 above)
Comment (sidvishnoi) [reply src/background/services/openPayments.ts:490]: This is something I'm leaving for the other PR, when we'd have logic like switchGrant and "top up" one time over existing grant. As of this PR, we'd only have one active grant at a time.

File: src/background/services/background.ts (hunk 6 above)
Comment (sidvishnoi) [reply src/background/services/background.ts:188]: Was just avoiding adding newlines 😁

File: src/background/services/events.ts (hunk 7 above)
Comment (sidvishnoi) [reply src/background/services/events.ts:9]: Helpful with logging/metrics to know if users get into multiple errors states one after other. Also, we can add transitions between states in popup/icons.

File: src/background/services/openPayments.ts (hunk 11 above)
Comment (raducristianpopa) [reply src/background/services/openPayments.ts:490]: > This is something I'm leaving for the other PR, when we'd have logic like switchGrant and "top up" one time over existing grant. As of this PR, we'd only have one active grant at a time.

Add a comment for it.

File: src/background/services/storage.ts [hunk 12]
```diff
@@ -1,16 +1,30 @@
...
+  StorageKey,
+  WalletAmount
+} from '@/shared/types'
 import { type Browser } from 'webextension-polyfill'
 import { EventsService } from './events'
 
 const defaultStorage = {
+  /**
+   * For migrations, increase this version and add a migration script in
+   * {@linkcode MIGRATIONS} below. New additions to structure that can be
```
Comment (sidvishnoi) [inline src/background/services/storage.ts:13]: Aside: In VSCode, MIGRATIONS is linked (can be followed on click)
image

**Review by dianafulga — APPROVED** [review]

**Comment by sidvishnoi** [comment]
@raducristianpopa @dianafulga In open outgoingPayment, we actually get `grantSpentDebitAmount` instead of balance. Should we compute balance on each outgoing payment, or just store spentAmount instead and compute as needed?

File: src/background/services/storage.ts [hunk 13]
```diff
@@ -31,10 +32,25 @@ const defaultStorage = {
...
+
   constructor(
     private browser: Browser,
     private events: EventsService
-  ) {}
+  ) {
+    const bigIntMin = (a: string, b: string) => (BigInt(a) < BigInt(b) ? a : b)
+    this.setRemainingBalanceRecurring = new DebounceWithQueue(
+      (amount: string) => this.set({ recurringGrantRemainingBalance: amount }),
+      (args) => [args.reduce((m, [e]) => (!m ? e : bigIntMin(m, e)), '')],
```
Comment (sidvishnoi) [inline src/background/services/storage.ts:45]: If we're getting multiple calls to `Storage.setRemainingBalance` (which maybe out of order depending on which createOutgoingPayment resolves first), we can collect the calls and update storage less often.
It might be ok to update balance in storage on each createOutgoingPayment also - maybe just unnecessary optimization, but I think more meaningful when we have like 5-6+ wallet addresses per page.

**Comment by dianafulga** [comment]
> @raducristianpopa @dianafulga In open outgoingPayment, we actually get `grantSpentDebitAmount` instead of balance. Should we compute balance on each outgoing payment, or just store spentAmount instead and compute as needed?

From what I know, we shouldn't compute anything on our side, it Should come from the API. Wait until it's implemented.

**Comment by sidvishnoi** [comment]
It appears it was decided in https://github.com/interledger/open-payments/issues/469 that they'd return spent amount only instead of balance (see edit in issue description).

So, now it's upto us how to use it. We already have what amount is approved in grant. Now, we can either subtract it on each storage, or only when we need the balance. I think if we don't need to know balance to know if we can make a payment (like before each payment or after, to reach the `out_of_founds` state), we can store spentAmount and compute balance as needed (for UI only I believe).

File: src/shared/helpers.ts [hunk 14]
```diff
@@ -88,6 +88,48 @@ export function debounceAsync<T extends unknown[], R extends Promise<unknown>>(
...
+ *   (total: number) => saveToStorage(total),
+ *   (collectedArgs) => collectedArgs.reduce(total, [val] => total + val, 0),
+ *   wait
+ * )
+ * debounceWithQueue.enqueue(10)
+ * debounceWithQueue.enqueue(15)
+ * // results in saveToStorage(25)
+ * ```
+ */
+export class DebounceWithQueue<Args extends unknown[]> {
```
Comment (raducristianpopa) [inline src/shared/helpers.ts:108]: What happens if the user closes the browser and we still have items in the queue?

Comment (sidvishnoi) [reply src/shared/helpers.ts:108]: They may get lost (upto a second of queue), but we can get latest data on next payment.

Anyway, I was testing this out and will move it into a separate PR when we store spent/balances on payment.

**Review by raducristianpopa — APPROVED** [review]
